### ISSUE TITLE
Packages resolved relative to module, not to calling script?

### ISSUE THREAD

**Issue by stevedonovan** [FascinatedBox/lily#267]
Here is the sample layout:

```
├── packages
│   ├── frodo
│   │   └── frodo.lly
│   └── test
│       └── test.lly
```
Any script in the dir containing `packages` can `use test` or `use frodo`. 

But `frodo.lly` cannot `use test`:

```
SyntaxError: Cannot load package 'test':
    no built-in package 'test'.
    packages/frodo/packages/test/test.lly
    packages/frodo/packages/test/test.so
    from packages/frodo/frodo.lly:1
```
Would it not make more sense to 'root' the package in the directory of the original calling script?

There's a workaround, but it isn't so pretty: symlinks.

**Comment by jesserayadkins** [comment]
The idea that I had is that each package forms its own sort of world. Always down, but never across. I hesitate to allow searching across, unless possibly as a fallback for searching down. There's a question of 'if frodo contains test, and also has test beside it, which should it pick'. If it searches across first, then that could break packages (you install a new test, and frodo isn't ready).

**Comment by stevedonovan** [comment]
As a fallback would make sense. Because non-trivial packages are composed of other packages and so forth - and don't have exclusive need for them. For instance, I have a lexer that needs xstr, and a JSON parser that needs lexer.  The symlink hacking would lead to very untidy trees, unless that's considered the price to pay for isolation.

**Comment by stevedonovan** [comment]
In a way, it makes `import foo` more flexible and useful than `use foo` which might be an undesired side-effect.

**Comment by jesserayadkins** [comment]
I don't know about making any changes. The current strategy's isolation is nice for being able to setup, say, travis on each package and knowing that each isolated package can be tested apart from others. I like that there's a divide between import and use: One for loading for the current package, and one for loading other packages.

I'll at best need a while to think it over, or might revisit it later.

**Comment by stevedonovan** [comment]
Fair point about testing in isolation, but the consequence is going to be a fairly convoluted package tree when a program needs packages that needs packages.  Either symlinks or lots of copies.  Yes, I know, tools can do this for us, but perhaps I'm still thinking of a global packages tree. Currently there is an asymmetry between built-in packages and external packages; the former are available everywhere. I write a lot of scripts (usually Lua these days) and the Lily approach  means that all scripts must live in their own directory, where they can see the shared packages.  I'm not sure there's a strong either/or distinction here basically.

**Comment by jesserayadkins** [comment]
I gave this some thinking over. The isolation of testing is something I want to keep. The general isolation of one interpreter versus another is another important feature, when I think about it. So I want to avoid an approach of setting an environment variable. I'm imagining someone embedding an interpreter, and not enjoying that kind of surprise.

At the same time, I like some of the points you make:

One, that there's a value in having a pool to pick from (without specifying an order). A pool can mitigate the fact that otherwise you'd have modules that descend very far down.

To add to that, there's hopefully going to be others like yourself who will end up maintaining multiple modules that are going to interdepend on each other. Being able to define a pool to pick from is useful there. But it's also useful for someone who wants to ship a full product: Said person can define a pool in a toplevel directory, and get a much flatter namespace.

Second, that import and use is starting to feel like it doesn't add much value. That 'use' is a distinction that doesn't offer much value. Roughly the same behavior can be achieved by having import look for `name/name.*` instead of having 'use' do packaging lookups. It cuts down on a layer (no more packages directory), and also feels easier. If I call to import 'x', and there exists 'x/x.lly', then that can be loaded with 'x/x.lly' acting as the root.

I'd like to avoid a magic-named __init__.py kind of name, because that doesn't feel quite right. Inference seems like it works here. There's a question that Python has (and which Lily would have, following such a change), which is 'what happens when there's a file 'x.lly' and a folder 'x' with 'x.lly' inside. The answer could be "don't do that", and that the first wins (file imports being more common than non-file imports).

I enjoy tangents, I've found. So I've come away with the following ideas:

One, that I'm quite particular on doing, is to see 'use' get folded into 'import'. Garden gets a small update. I don't have tests for packages, and removing 'use' is I expect going to be moderately easy. The more I look at it, the less I see the value of making a distinction between use and import, when import can have that special path-ing that use has.

Another idea is to allow a flag to be passed that specifies a pool directory. You could then use this in tests, specifying that there's a pool directory to pull folders from: `pool/x/x.lly` type searches (or library paths.

Another thought is to make the root directory of the first file loaded to automatically be the pool, but to give it a lower precedence than lower package lookups. I suspect what you're wanting is a flat directory of packages expanding from it, so that you can run tests without needing to setup/teardown symlinks. So I'm not sure how much of a help it would be for that.

**Comment by stevedonovan** [comment]
I have to agree that `use` distinct from `import` is a point of confusion - yet another thing to explain!  I think the less concepts you have to transfer, the easier it is for someone to pick up a new language. But yes, we do need control - for testing, for sandboxing and so forth. `x/x.lly` feels better.

If the main module (`__main__`) gets special treatment, then the 'sprawling tree' problem goes away - I actually don't mind a program having its own tree, it's every package having its own tree that seemed awkward ;)

**Comment by FascinatedBox** [comment]
Commit 1b080e1 (which I pointed toward the wrong issue) gets rid of `use`, and merges the effective behavior into `import`. I rewrote some of the path selection so that there's a packages directory that's rooted from the base commit. I'm thinking of, in the future, allowing extra search paths to be added (the new 'algorithm' uses transformed paths like Lua's does), so it wouldn't be a stretch. This should work now, but give me a poke if it doesn't. :)
